pippy/PipelineSchedule.py

```diff
@@ -603,27 +603,22 @@ def step_microbatches(
         losses: Optional[List] = None,
     ):
         """
-        # n_loop = n_stage / n_pp
-        # run microbatches in sequences of NPp
+        Operate on the microbatches for interleaved 1f1b schedule (https://arxiv.org/pdf/2104.04473.pdf).
 
-        schedule operates at the rank level
-
-        highest rank has a warmup (F only) count of [len(stages) - 1] * seq_size
-        each hop away from highest rank adds 2 warmup stages
+        Highest rank has a warmup (F only) count of [len(stages) - 1] * group_size
+        and each hop away from highest rank adds 2 warmup stages due to:
```

Comment:
What does `group_size` mean here? Number of PP ranks? If so, is `[len(stages) - 1] * group_size` a square of # PP ranks?
And what does "hop" mean?

Reply:
Yep `group_size` is pp ranks. And the warmup steps is going to be a multiple of the pp ranks. Hop means for every rank less than the highest. The comment is a bit old so i will update it.

Reply:
oh, so `len(stages)` means the number of **local** stages?